### PR TITLE
CSS3 transitions for all fade animations

### LINKED ISSUE #37691: FlexSlider fade animations don't need to use $.animate()
### Describe the solution you'd like

When configuring **FlexSlider** to use the _fade_ animation type (which is NOT the WooCommerce default, that would be the _slide_ animation type), **jquery.flexslider.js** uses CSS3 transitions on touch-enabled devices; and `$.animate()` otherwise.

Those `$.animate()` calls are simple opacity changes, they can be easily replaced with CSS3 transitions similar to the ones **jquery.flexslider.js** already uses on touch-enabled devices. Thus, we can reduce **FlexSlider's** dependency on **jQuery**.

### Describe alternatives you've considered

_No response_

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by apenchev**
### Submission Review Guidelines:

- I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
- I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
- I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/). 
- Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.

### Changes proposed in this Pull Request:

Closes #37691.

### How to test the changes in this Pull Request:

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Add this custom code to your theme's `functions.php` in order to enable fade animations:
```
function carousel_options( $options ) {
	return array_merge(
		$options,
		array(
			'animation' => 'fade',
			'smoothHeight' => false, // smoothHeight is extremely buggy with fade for some reason
			'animationSpeed' => 2600,
			'fadeFirstSlide' => false,
		)
	);
}
add_filter( 'woocommerce_single_product_carousel_options', 'carousel_options' );
```
2. Visit a Single Product Page on a non-touch-enabled device (a PC or a laptop).
3. Confirm that the first gallery image does NOT fadeIn on page load.
4. Confirm that switching between gallery images happens with a fade animation that lasts 2.6 seconds.
5. Enable `fadeFirstSlide` in the custom code in your theme's `functions.php`:
```
function carousel_options( $options ) {
	return array_merge(
		$options,
		array(
			'animation' => 'fade',
			'smoothHeight' => false, // smoothHeight is extremely buggy with fade for some reason
			'animationSpeed' => 2600,
			'fadeFirstSlide' => true,
		)
	);
}
add_filter( 'woocommerce_single_product_carousel_options', 'carousel_options' );
```
6. Visit a Single Product Page on a non-touch-enabled device (a PC or a laptop).
7. Confirm that the first gallery image DOES fadeIn on page load.

`'smoothHeight' => false,` is needed since `smoothHeight` doesn't play nice with `fade` - that's an existing bug with FlexSlider.

### CODE REVIEW

**Comment by apenchev** [comment]
@vedanshujain Would you be able to take a look?

**Comment by apenchev** [comment]
:/

**Comment by apenchev** [comment]
> @vedanshujain Would you be able to take a look?

**Review by Konamiman — APPROVED** [review]

**Comment by Konamiman** [comment]
Testing works as expected, but I'm requesting analysis since I'm not really expert in JavaScript, thanks.

**Comment by alopezari** [comment]
Thanks @Konamiman!

Will make sure to test this thoroughly once it gets included in the next release package 👍

File: plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js
Comment (gigitux) [inline plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js]: Is there a reason for this change? Could it be an issue for "extensions" that add a handler on the transitionend?

**Review by gigitux — COMMENTED** [review]
Thanks for working on this! 

I've thoroughly tested the pull request and it looks good to me. However, my primary concern is to ensure that this PR won't cause any regressions or disrupt existing plugins or themes. I've pinpointed a specific concern with line 728, and I'd like to hear your insights before granting my approval.

File: plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js
Comment (apenchev) [reply plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js]: Thanks for testing! Old code used to call `jQuery.animate()` and pass `slider.wrapup` as [the `complete` callback](https://api.jquery.com/animate/#animate-properties-duration-easing-complete-complete). In order to replicate [the `complete` callback from jQuery](https://api.jquery.com/animate/#animate-properties-duration-easing-complete-complete), we now call `slider.wrapup` during the CSS `transitionend`.

If there are plugins subscribing to `transitionend` on `slider.slides`, they are doing it knowing that the subscription would only work on touch-enabled devices. Remember, in the production code CSS transitions are ONLY used inside `if (touch) { }`.

In this PR, we remove all current `transitionend` subscriptions with `.off("transitionend")` inside `if (!touch) { }`.

Comment (gigitux) [reply plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js]: Hi, @apenchev - thanks for the explanation. Just to be sure. Given this:

> If there are existing plugins subscribing to transitionend on slider.slides, they are doing it knowing that the subscription would only work on touch-enabled devices (remember, before this PR, CSS transitions are ONLY used inside if (touch) { })

> inside of if (!touch) { } we remove existing transitionend subscriptions before adding our own transitionend subscription

It means that this PR doesn't introduce any potential breaking change. Right?

Comment (apenchev) [reply plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js]: Unfortunately, the PR does introduce a potential breaking change. My explaination was not correct, as I assumed that we do a .transitionend cleanup inside of `if (!touch) { }`, but we don't clean up everything.

I will try and send a small fix that attempts to do a full .transitionend cleanup inside of `if (!touch) { }`.

Comment (apenchev) [reply plugins/woocommerce/client/legacy/js/flexslider/jquery.flexslider.js]: I've pushed the commit with the proper transitionend cleanup.

**Review by gigitux — CHANGES_REQUESTED** [review]
Hi @apenchev,
Sorry for the late answer!
Thanks for your work in addressing the feedback and improving the code to avoid any potential breaking changes.

I noticed that with the latest changes, the thumbnails aren't visible anymore:


https://github.com/woocommerce/woocommerce/assets/4463174/f399e673-6238-4b5f-8df6-b52b2d81e98a

Can you check and see if you are able to replicate it?

**Comment by apenchev** [comment]
> 'animationSpeed' => 1600,

I can't reproduce it, looks good on my end. :/ I'm using the StoreFront default theme, WooCommerce status is:

WooCommerce: 8.3.1
WooCommerce REST API: 8.3.1
WooCommerce Blocks package: 11.4.9
Action Scheduler package: 3.6.4
WordPress: 6.4.1

**Comment by apenchev** [comment]
> Hi @apenchev, Sorry for the late answer! Thanks for your work in addressing the feedback and improving the code to avoid any potential breaking changes.
> 
> I noticed that with the latest changes, the thumbnails aren't visible anymore:
> 
>  Screen.Capture.Nov.14.mov 
> Can you check and see if you are able to replicate it?

Yes, the bug happens on the Twenty Twenty-Four theme when you set the animation to `fade` - whether you're using this PR or not. I am using the StoreFront theme for testing so I didn't see this bug.

This PR doesn't add this bug, the bug appears on the Twenty Twenty-Four theme even on a clean install simply by adding:

```
function carousel_options( $options ) {
	return array_merge(
		$options,
		array( 'animation' => 'fade' )
	);
}
add_filter( 'woocommerce_single_product_carousel_options', 'carousel_options' ); 
```

**Comment by apenchev** [comment]
@gigitux I've greatly simplified the code in the PR and re-tested it. It still cleans up all `transitionend` subscriptions inside `if (!touch) { }` in order to make sure that we don't introduce regressions, as we discussed.

**Review by gigitux — COMMENTED** [review]
Thanks for your effort on this PR! It is LGTM and it is ready to be shipped. Before approving the PR, can you rebase it with trunk?

**Comment by apenchev** [comment]
trunk merged.

**Review by gigitux — APPROVED** [review]
Thanks again for your effort! I'll merge the PR once all CI checks pass.

**Pull request merged**
